smoke-tests/src/test/java/com/splunk/opentelemetry/helper/windows/WindowsTestContainerManager.java

```diff
@@ -0,0 +1,524 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.helper.windows;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.command.CreateContainerCmd;
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.command.PullImageResultCallback;
+import com.github.dockerjava.api.exception.NotFoundException;
+import com.github.dockerjava.api.model.ExposedPort;
+import com.github.dockerjava.api.model.HostConfig;
+import com.github.dockerjava.api.model.PortBinding;
+import com.github.dockerjava.api.model.Ports;
+import com.github.dockerjava.core.DefaultDockerClientConfig;
+import com.github.dockerjava.core.DockerClientImpl;
+import com.github.dockerjava.httpclient5.ApacheDockerHttpClient;
+import com.splunk.opentelemetry.helper.AbstractTestContainerManager;
+import com.splunk.opentelemetry.helper.TargetWaitStrategy;
+import com.splunk.opentelemetry.helper.TestImage;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.Socket;
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.regex.Pattern;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream;
+import org.apache.commons.io.IOUtils;
+import org.rnorth.ducttape.TimeoutException;
+import org.rnorth.ducttape.ratelimits.RateLimiter;
+import org.rnorth.ducttape.ratelimits.RateLimiterBuilder;
+import org.rnorth.ducttape.unreliables.Unreliables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class WindowsTestContainerManager extends AbstractTestContainerManager {
```

Comment:
Are this class and `LinuxTestContainerManager` functionally equivalent? Or is `LinuxTestContainerManager` more powerful?

Reply:
@agoallikmaa ^^ ?

Reply:
They effectively perform the same task, but `LinuxTestContainerManager` uses `testcontainers`, while `WindowsTestContainerManager` uses `docker-java` directly, as `testcontainers` does not support Windows containers. Another difference is obviously that `WindowsTestContainerManager` uses Windows collector and fake backend images while `LinuxTestContainerManager` uses Linux ones.

Reply:
My point here is we use cannot use testcontainers on both platforms and thus brought `docker-java`, can we still have one `ContainerManger` for both platforms? Or do we benefit somehow from using testcontainers for Linux? Currently seems like duplicated effort.

Reply:
There are actually some nice-to-haves in the testcontainers implementation that I did not duplicate as that would not have been worth it just for this one use case - such as containers being more reliably cleaned up without relying on the stop/kill being called from Java code.

Also, Testcontainers would be the preferred way to do this, and hopefully they'll get around to supporting Windows images. And if they do, then the `docker-java` implementation could be thrown away and replaced with a few ifs for Windows.

Ironically this cleanup mechanism with an auxiliary container is what's blocking them from supporting Windows, as they seem to be stubborn about this container always having to be a Linux container, even though what runs on it is written in Go which would work just as nicely in a Windows container.

Reply:
There are actually some nice-to-haves in the Testcontainers implementation that I did not implement in the docker-java one as that would not have been worth it just for this one use case. The main one is that it makes sure the containers are cleaned up without relying on the stop/kill being called from Java code.

Also, Testcontainers would be the preferred way to do this, so ideally this docker-java custom implementation could be removed once they add support for Windows images.